[docs] update web publishing docs (#25934)
expo/expo: docs/pages/distribution/publishing-websites.mdx

description: Different ways to publish the Expo web app with third-party services.
---

import { APIBox } from '~/components/plugins/APIBox';
import { Step } from '~/ui/components/Step';
import { Tab, Tabs, TabsGroup } from '~/ui/components/Tabs';
import { Table, TableHead, Row, HeaderCell } from '~/ui/components/Table';
import { Terminal } from '~/ui/components/Snippet';
import { YesIcon, NoIcon, PendingIcon } from '~/ui/components/DocIcons';

A web app created using Expo can be served locally for testing out the production behavior. Once the testing phase checks out, you can choose from a variety of third-party services to host it.

## Output targets

> Expo supports three output targets for web apps. The output target can be configured in the **app.json** file:

Reviewer: ```suggestion
Expo Router supports three output targets for web apps. The `web.output` target can be configured in the **app.json** file:
```

> Expo Router supports three output targets for web apps. The `web.output` target can be configured in the **app.json** file:

Reviewer 2: ```suggestion
The [`web.output`](/versions/latest/config/app/#output) target can be configured in the [app config](/workflow/configuration/) to set the export method for the web app:
```

Let's add link to the `web.output` config definition and also let's refer app config instead of `app.json`. It's what we use for consistency. Let's also briefly explain what it does.

```json app.json
{
  "expo": {
    "web": {
      "output": "server",
      "bundler": "metro"
    }
  }
}
```

> > If you are using Expo Web without Expo Router, only `single` is supported

Reviewer: ```suggestion
> If you are using legacy Expo Webpack (deprecated), only `single` is supported.
```

> > If you are using legacy Expo Webpack (deprecated), only `single` is supported.

Reviewer 2: ```suggestion
```

We can combine the info in this callout with webpack info later within this section. Having two contexts for the same information above and below the table didn't make sense to me.

> 

Reviewer 2: ```suggestion

Expo Router supports three output targets for web apps.

```

> | Output             | Expo Router | API Routes  |  Description                                                                                                                   |

Author: I would like advise on this table. I'm trying to explain the outputs, while also showing the API routes only works with 1 output, while also explaining that these are Expo Router only, and Metro/Webpack don't have these features.

Its got a lot going on

Reviewer 2: I think the table looks fine. To add more context of what you are trying to explain, let's add it above this table.

| ------------------ | ----------- | ----------- |  ----------------------------------------------------------------------------------------------------------------------------- |

> | `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA), with a single index.html in the output folder, and has no statically indexable HTML. |

Reviewer 2: ```suggestion
| `single` (default) | <YesIcon /> | <NoIcon />  |  Outputs a Single Page Application (SPA) with a single **index.html** in the output folder and has no statically indexable HTML. |
```

Let's remove the commas and highlight the file name in bold per our writing style guide.

> | `server`           | <YesIcon /> | <YesIcon /> |  Creates a client & server folder. Client files are outputed as seperate HTML files. API routes as seperate JavaScript files   |

Reviewer 2: ```suggestion
| `server`           | <YesIcon /> | <YesIcon /> |  Creates **client** and **server** directories. Client files are output as separate HTML files. API routes as separate JavaScript files for hosting with a custom Node.js server.   |
```

Added a bit more context for the API routes part of this description based on https://docs.expo.dev/versions/latest/config/app/#output. It wasn't clear what would one do with API routes here or why they should know about it.

> | `static`           | <YesIcon /> | <NoIcon />  |  Outputs seperate HTML files for every route in the app/ directory                                                             |

Reviewer 2: ```suggestion
| `static`           | <YesIcon /> | <NoIcon />  |  Outputs separate HTML files for every route in the **app** directory.                                                             |
```

> > **warning** In SDK 50+, publishing with `webpack` is deprecated

Reviewer 2: ```suggestion
> **warning** In SDK 50 and above, publishing with `webpack` is deprecated. 
```

> Expo only supports the single output format when using `webpack`.

Reviewer 2: ```suggestion
Expo only supports the `single` output format using `webpack`.
```

```json app.json
{
  "expo": {
    "web": {
      "bundler": "webpack"
    }
  }
}
```

| Bundler   | Expo Router | API Routes | Description                                                                             |
| --------- | ----------- | ---------- | --------------------------------------------------------------------------------------- |

> | `webpack` | <NoIcon />  | <NoIcon /> | Outputs a Single Page Application (SPA), with a single index.html in the output folder. |

Reviewer 2: ```suggestion
| `webpack` | <NoIcon />  | <NoIcon /> | Outputs a Single Page Application (SPA) with a single **index.html** in the output folder. |
```

<Tabs>

<Tab label="single">

<Terminal cmd={['$ npx serve dist --single']} />

Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will not work.

</Tab>

<Tab label="static">

<Terminal cmd={['$ npx serve dist']} />

Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will work.

> </Tab>

Reviewer 2: Let's move this after the server since it seems to be the pattern for this whole document (and we should follow it even though it might give our users a tiny but uniformed experience).

<Tabs>

<Tab label="single">

<Terminal cmd={['$ npx serve dist --single']} />

> Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features won't work.

Reviewer 2: ```suggestion
Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will not work.
```

Let's avoid the contraction here.

</Tab>

<Tab label="static">

<Terminal cmd={['$ npx serve dist']} />

> Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features won't work.

Reviewer 2: ```suggestion
Open [`http://localhost:5000`](http://localhost:5000) to see your project in action. This method is **HTTP only**, so permissions, camera, location, and many other secure features will work.
```

I just realized this is outside the scope of this PR so feel free to ignore this and the other similar suggestion.

</Tab>

<Tab label="server">

> There is no prebuilt solution. You will need to create your own server to serve the static files and API routes.

Author: This statement feels bad. How can I improve the language here? I tried to get `npx serve` working with this output but I gave up.

@kitten Got any advise?
